doc/sphinx/addendum/extraction.rst

```diff
@@ -399,6 +399,109 @@ As an example of translation to a non-inductive datatype, let's turn
 
    Extract Inductive nat => int [ "0" "succ" ] "(fun fO fS n -> if n=0 then fO () else fS (n-1))".
 
+Generating FFI Code
+~~~~~~~~~~~~~~~~~~~
+
+The plugin provides mechanisms to generate (currently only) OCaml code to
+interface the generated OCaml code with C programs. In order to link compiled
+OCaml code with C code, the linker needs to know
+
+ * which C functions will be called by the ML code (externals)
+ * which ML functions shall be accessible by the C code (callbacks)
+
+.. cmd:: Extract External Constant @qualid => @string
+
+   Same as the :cmd:`Extract Constant`, except that the given ML terms
+   will not be declared via ``let`` but in the form
+   ``external @qualid : ML type = "@string"``.
+   Also, no ``val`` declaration will be generated in the respective
+   mli file.
+
+.. caution::
+
+   * The external function name provided by the user as :n:`@string`
+     is currently **not** checked at all by extraction, even for syntax errors.
+
+   * It is the responsibility of the user to ensure that the C
+     functions given to realize the axioms do have the expected or compatible
+     types. In fact, the strings containing realizing code are just copied to the
+     extracted files.
+
+   * A previously used :cmd:`Extract Constant` for :n:`@qualid` will be overridden
+     by the command :cmd:`Extract External Constant`, although internally, the
+     specified :n:`@qualid` will still be part of the (inline) constant extraction
+     table. This is by design, since, when defining a function to be external, a
+     ``val`` declaration and ``let`` definition would collide with the external declaration
+     during compilation. Also, an external function cannot be inlined. Thus, the
+     command :cmd:`Extract Inlined Constant` has to be overridden, too.
+
+.. note::
+   Since the generated external expression contains the mapped OCaml type, the
+   C/C++ compiler is able to determine whether the ML function and respective
+   C/C++ function match in the arity and whether the types are compatible. Thus,
+   the type safety is not undermined when using :cmd:`Extract External Constant`.
+
+.. exn:: Extract External Constant is supported only for OCaml extraction
+
+   Calling foreign functions is gravely different between OCaml, Haskell and
+   Scheme. Thus, extracting externals does not make sense when extracting into
+   anything other than OCaml.
```

Comment:
The fact that FFI is different between OCaml and Haskell and Scheme does not mean than it does not make sense to extract externals to Haskell or Scheme. In fact, the difference between these systems is almost negligible. In OCaml, you do `external name : type = "string"`, while in Haskell you would do `foreign import capi "string" name :: type`, and in Scheme you would do `define name (foreign-procedure "string" type)`.

By the way, this makes the whole "external" naming ill-suited for the feature. It would be better to have a more agnostic naming, e.g., `Extract Foreign Constant`.

Reply:
This is also correct. Again, sorry for that one. I read the documentation of Haskell and Scheme concerning FFIs. This should be possible with quite few changes to my code. For Haskell, I need an additional line at the begin of the file. Should I extend the support do Haskell and Scheme or restrict my changes to OCaml for now. Also I saw, that Haskell has `foreign export` which is comparable to the OCaml callbacks. But I don't see a well-fitting common term for those concepts - except from Extract Closure. But the term closure is quite overloaded.
For Scheme, I have to dig through the documentation concerning something comparable to OCaml callback registration.

Also, I like the proposal `Extract Foreign Constant` and adopted that.